cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,248 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.framework;
+
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import co.cask.cdap.logging.write.FileMetaDataManager;
+import com.google.common.base.Throwables;
+import com.google.common.io.Closeables;
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileWriter;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class including logic for getting avro files to write to. Used by {@link CDAPLogAppender}
+ */
+class AvroFileManager {
+  private static final Logger LOG = LoggerFactory.getLogger(AvroFileManager.class);
+
+  private final long maxLifetimeMillis;
+  private final Map<LogPathIdentifier, LogFileOutputStream> fileMap;
+  private final Location logsDirectoryLocation;
+  private final FileMetaDataManager fileMetaDataManager;
+  private final int syncIntervalBytes;
+  private final Schema schema;
+
+  AvroFileManager(long maxFileLifetimeMs, int syncIntervalBytes, Schema schema,
+                  FileMetaDataManager fileMetaDataManager,
+                  LocationFactory locationFactory) {
+    this.maxLifetimeMillis = maxFileLifetimeMs;
+    this.syncIntervalBytes = syncIntervalBytes;
+    this.schema = schema;
+    this.fileMetaDataManager = fileMetaDataManager;
+    this.logsDirectoryLocation = locationFactory.create("logs");
+    this.fileMap = new HashMap<>();
+  }
+
+  /**
+   * Get an avro file for the give logging context and timestamp - return an already open file,
+   * or create and return a new file or rotate a file based on time and return.
+   * @param logPathIdentifier
+   * @param timestamp
+   * @return LogFileOutputStream
+   * @throws IOException
+   */
+  public LogFileOutputStream getAvroFile(LogPathIdentifier logPathIdentifier, long timestamp) throws IOException {
+    LogFileOutputStream logFileOutputStream = fileMap.get(logPathIdentifier);
+    // If the file is not open then set reference to null so that a new one gets created
+    if (logFileOutputStream == null) {
+      logFileOutputStream = createAvroFile(logPathIdentifier, timestamp);
+    }
+    // rotate the file if needed (time has passed)
+    logFileOutputStream = rotateFile(logFileOutputStream, logPathIdentifier, timestamp);
```

Comment:
Shouldn't it in the `else` part since you don't need to rotate if the stream is newly created?